vsphere/datadog_checks/vsphere/api_rest.py

```diff
@@ -0,0 +1,122 @@
+# (C) Datadog, Inc. 2019-present
+# All rights reserved
+# Licensed under Simplified BSD License (see LICENSE)
+from collections import defaultdict
+
+import requests
+from pyVmomi import vim
+from vmware.vapi.vsphere.client import create_vsphere_client
+
+from .api import APIConnectionError, smart_retry
+
+MOR_TYPE_MAPPING = {
+    'HostSystem': vim.HostSystem,
+    'VirtualMachine': vim.VirtualMachine,
+    'Datacenter': vim.Datacenter,
+    'Datastore': vim.Datastore,
+    'ClusterComputeResource': vim.ClusterComputeResource,
+}
+
+
+class VSphereRestAPI(object):
+    """
+    Abstraction class over the vSphere REST api using the vsphere-automation-sdk-python library
+    """
+
+    def __init__(self, config, log):
+        self.config = config
+        self.log = log
+
+        self._client = None
+        self.smart_connect()
+
+    def smart_connect(self):
+        """
+        Create a vSphere client.
+        """
+        session = requests.Session()
+        session.verify = self.config.ssl_verify
+        session.cert = self.config.ssl_capath
+
+        try:
+            client = create_vsphere_client(
+                session=session,
+                username=self.config.username,
+                password=self.config.password,
+                server=self.config.hostname,
+            )
+        except Exception as e:
+            err_msg = "Connection to {} failed, hostname: {}".format(self.config.hostname, e)
+            raise APIConnectionError(err_msg)
+
+        self._client = client
+
+    @smart_retry
+    def get_resource_tags(self):
+        """
+        Get resource tags.
+
+        Response structure:
+
+            {
+                <RESOURCE_TYPE>: {
+                    <RESOURCE_MOR_ID>: ['<CATEGORY_NAME>:<TAG_NAME>', ...]
+                },
+                ...
+            }
+        """
+        categories = self._get_categories()
+        tags = self._get_tags(categories)
+        tag_ids = list(tags.keys())
+        tag_associations = self._get_tag_associations(tag_ids)
+
+        # Initialise resource_tags
+        resource_tags = {resource_type: defaultdict(list) for resource_type in MOR_TYPE_MAPPING.values()}
+
+        for tag_asso in tag_associations:
+            tag = tags[tag_asso.tag_id]
+            for resource_asso in tag_asso.object_ids:
+                resource_type = MOR_TYPE_MAPPING.get(resource_asso.type)
+                if not resource_type:
+                    self.log.debug(
```

Comment:
I expect this log message to be emitted pretty often for a normal use case. There are other resource types that you can attach tags to.

Maybe a single log line at the end of the loop that tells if tags were found for each resource type would be better.

Reply:
You are right, it's expected. There are actually more types of resources that can be tagged than those in `MOR_TYPE_MAPPING`.

It would be great if we could fetch the tag associations for specific types, but seems not possible.

I'm replacing with a different debug logs.